### ISSUE TITLE
The KISS principle in machine learning

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3433]
The KISS principle in machine learning<br>
<br>
http://bit.ly/2KLB93P<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div> <p>Software developers are already familiar with the <a href="https://en.wikipedia.org/wiki/KISS_principle">KISS principle</a>. Roughly speaking, it refers to the old wisdom that simple solutions are to be preferred, while <em>unnecessary</em> complexity should be avoided. In machine learning, this means that one should increase the complexity iteratively, starting from simple models and—if necessary—working upwards to more complicated ones. I was recently reminded and humbled by the wisdom of KISS while working on a novel topology-based kernel for machine learning.</p> <h1 id="prelude-a-paper-on-deep-graph-kernels">Prelude: A Paper on Deep Graph Kernels</h1> <p>A <a href="https://users.soe.ucsc.edu/~vishy/pubs/YanVis15.pdf">recent paper by Yanardag and Vishwanathan</a> introduced a way of combining graph kernels with deep learning techniques. The new framework basically yields a way of using graph kernels, such as <a href="http://proceedings.mlr.press/v5/shervashidze09a/shervashidze09a.pdf"><em>graphlet kernels</em></a> in a regular deep learning setting.</p> <p>The two researchers used some interesting data sets for their publication. Most notably, they extracted a set of co-occurrence networks (more about that in a minute) from <a href="https://www.reddit.com">Reddit</a>, a content aggregation and discussion site. Reddit consists of different communities, the <em>subreddits</em>. Each subreddit deals with a different topic, ranging from archaeology to zoology. The posting style of these subreddits varies a lot. There are several subreddits that are based on a question–answer format, while others are more centred around individual discussions.</p> <p>Yanardag and Vishwanathan hence crawled the top submissions from the subreddits <a href="https://www.reddit.com/r/IAmA"><em>IamA</em></a>, <a href="https://www.reddit.com/r/AskReddit"><em>AskReddit</em></a>, both of which are based on questions and answers, as well as from <a href="https://www.reddit.com/r/TrollXChromosomes"><em>TrollXChromosomes</em></a>, and <a href="https://www.reddit.com/r/atheism"><em>atheism</em></a>, which are discussion-based subreddits. From every submission, a graph was created by taking all the commenters of a thread as nodes and connecting two nodes by an edge if one user responds to the comment of another user. We can see that this is an extremely simple model—it represents only a fraction of the information available in every discussion. Nonetheless, there is some hope that qualitatively different behaviours will emerge. More precisely, the assumption of Yanardag and Vishwanathan is that there is some quantifiable difference between question–answer subreddits and discussion-based subreddits. Their paper aims to <em>learn</em> the correct classification for each thread. Hence, given a graph, we want to teach the computer to tell us whether the graph is more likely to arise from a question–answer subreddit or from a discussion-based one.</p> <p>The two researchers refer to this data set as <em>REDDIT-BINARY</em>. It is now available in <a href="https://ls11-www.cs.tu-dortmund.de/staff/morris/graphkerneldatasets">a repository of benchmark data sets for graph kernels</a>, gracefully provided and lovingly curated by the <a href="https://ls11-www.cs.tu-dortmund.de/start">CS department of Dortmund University</a>.</p> <h1 id="interlude-looking-at-the-data">Interlude: Looking at the Data</h1> <p>Prior to actually <em>working</em> with the data, let us first take a look at it in order to get a feel for its inherent patterns. To this end, I used <a href="https://github.com/Submanifold/Aleph">Aleph, my library for topological data analysis</a> read and convert the individual graphs of every discussion to the simpler <a href="https://gephi.org/users/supported-graph-formats/gml-format">GML format</a>. See below if you are also interested in the data. Next, I used <a href="https://gephi.org">Gephi</a> to obtain a force-directed visualization of some of the networks.</p> <p>A look at a question–answer subreddit shows a central <em>core</em> structure of the data set, from which numerous strands—each corresponding most probably to smaller discussions—emerge:</p> <figure><a href="http://bastian.rieck.me/images/graph_qa_subreddit.png"> <img src="http://bastian.rieck.me/images/graph_qa_subreddit.png" alt="A graph visualization of Q&amp;A subreddits"></a> </figure><p>The discussion-based subreddit graph, on the other hand, exhibits a larger <em>depth</em>, manifesting themselves in a few long strands. Nonetheless, a similar central core structure is observable as well.</p> <figure><a href="http://bastian.rieck.me/images/graph_discussion_subreddit.png"> <img src="http://bastian.rieck.me/images/graph_discussion_subreddit.png" alt="A graph visualization of discussion-based subreddits"></a> </figure><p>Keep in mind that the selected examples are not necessarily representative—I merely picked two of the large graphs in the data set to obtain an idea of how the data looks.</p> <h1 id="a-complex-classification">A Complex Classification</h1> <p>A straightforward way to classify those networks would be to use graph kernels, such as the <a href="http://proceedings.mlr.press/v5/shervashidze09a/shervashidze09a.pdf"><em>graphlet kernels</em></a>. The basic idea behind these kernels is to measure a dissimilarity between two graphs by means of, for example, the presence or absence of certain subgraphs. At least this is the strategy pursued by the <em>graphlet kernel</em>. Other kernels may instead opt for comparing <em><a href="https://en.wikipedia.org/wiki/Random_walk">“random walks”</a></em> on both graphs. A common theme of these kernels is that they are rather expensive to compute. In many applications, they are the only hope of obtaining suitable dissimilarity information without having to solve the graph isomorphism problem, which is even more computationally expensive. Hence, graph kernels are often the only suitable way of assessing the dissimilarity between graphs. They are quite flexible and, as the authors show, can even be integrated into a deep learning framework.</p> <p>As for their performance, Yanardag and Vishwanathan report an accuracy of 77.34% (standard deviation 0.18) for graphlet kernels and 78.04% (standard deviation 0.39) for <em>deep</em> graphlet kernels, which is very good considering the baseline for random guessing is 50%.</p> <h1 id="a-simple-classification">A Simple Classification</h1> <p>In line with the KISS principle, I wanted to figure out alternative ways of achieving similar accuracy values with a better performance, if possible. This suggests a feature-based approach with features that can be calculated easily. There are numerous potential options for choosing these features. I figured that good choices include the <a href="https://en.wikipedia.org/wiki/Clustering_coefficient">“average clustering coefficient”</a> of the graph, the <a href="https://en.wikipedia.org/wiki/Degree_(graph_theory)">“average degree”</a>, the average shortest path length, the <a href="https://en.wikipedia.org/wiki/Dense_graph">“density”</a>, and the <a href="https://en.wikipedia.org/wiki/Distance_(graph_theory)">“diameter”</a> of the graph. Among these, the average shortest path length and the diameter take longest to compute because they essentially have to enumerate all shortest paths in the graph. So I only used the remaining three features, all of which are computable in polynomial time.</p> <p>I used the excellent <a href="https://networkx.github.io">NetworkX package</a> for Python to do most of the heavy lifting. Reading a graph and calculating its features is as easy as it gets:</p> <div><pre><code><span>import</span> <span>networkx</span> <span>as</span> <span>nx</span> <span>G</span> <span>=</span> <span>nx</span><span>.</span><span>read_gml</span><span>(</span><span>filename</span><span>,</span> <span>label</span><span>=</span><span>'id'</span><span>)</span> <span>average_clustering_coefficient</span> <span>=</span> <span>nx</span><span>.</span><span>average_clustering</span><span>(</span><span>G</span><span>)</span> <span>average_degree</span> <span>=</span> <span>np</span><span>.</span><span>mean</span><span>(</span> <span>[</span><span>degree</span> <span>for</span> <span>_</span><span>,</span><span>degree</span> <span>in</span> <span>nx</span><span>.</span><span>degree</span><span>(</span><span>G</span><span>)</span> <span>]</span> <span>)</span> <span>density</span> <span>=</span> <span>nx</span><span>.</span><span>density</span><span>(</span><span>G</span><span>)</span></code></pre></div> <p>I collect these values in a <code>pandas.DataFrame</code> to simplify their handling. Now we have to choose some classifiers. I selected <a href="http://scikit-learn.org/stable/modules/tree.html"><em>decision trees</em></a>, <a href="http://scikit-learn.org/stable/modules/svm.html"><em>support vector machines</em></a>, and <a href="http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html"><em>logistic regression</em></a>. Next, let us train and test each of these classifiers by means of 10-fold <a href="https://en.wikipedia.org/wiki/Cross-validation_(statistics)">“cross-validation”</a>.</p> <div><pre><code><span>X</span> <span>=</span> <span>df</span><span>.</span><span>values</span> <span>y</span> <span>=</span> <span>labels</span> <span>classifiers</span> <span>=</span> <span>[</span> <span>DecisionTreeClassifier</span><span>(),</span> <span>LinearSVC</span><span>(),</span> <span>LogisticRegression</span><span>()</span> <span>]</span> <span>for</span> <span>clf</span> <span>in</span> <span>classifiers</span><span>:</span> <span>scores</span> <span>=</span> <span>cross_val_score</span><span>(</span><span>clf</span><span>,</span> <span>X</span><span>,</span> <span>y</span><span>,</span> <span>cv</span><span>=</span><span>10</span><span>)</span> <span>print</span><span>(</span><span>"Accuracy: </span><span>%0.4f</span><span> (+/- </span><span>%0.2f</span><span>)"</span> <span>%</span> <span>(</span><span>scores</span><span>.</span><span>mean</span><span>(),</span> <span>scores</span><span>.</span><span>std</span><span>()</span> <span>*</span> <span>2</span><span>))</span></code></pre></div> <p>As you can see, I am using the <em>default values</em> for every classifier—no grid search or other technique for finding better hyperparameters for now. Instead, I want to see how these classifiers perform out of the box.</p> <h1 id="astonishing-results">Astonishing Results</h1> <p>The initial test resulted in the following accuracy values: 77.84% (standard deviation 0.16) for <em>decision trees</em>, 64.14% (standard deviation 0.45) for <em>support vector machines</em>, and 55.54% (standard deviation 0.49) for <em>logistic regression</em>. At least the first result is highly astonishing—without any adjustments to the classifier <em>whatsoever</em>, we obtain a performance that is en par with more complex learning strategies! Recall that Yanardag and Vishwanathan adjusted the hyperparameters during training, while this approach merely used the default values.</p> <p>Let us thus focus on the decision tree classifier first and figure out <em>why</em> it performs the way it did. To this end, let us take a look at what the selected features look like and how important they are. To do this, I just added a single training instance, based on standard test–train split of the decision tree classifier:</p> <div><pre><code><span>X_train</span><span>,</span> <span>X_test</span><span>,</span> <span>y_train</span><span>,</span> <span>y_test</span> <span>=</span> <span>train_test_split</span><span>(</span><span>X</span><span>,</span><span>y</span><span>,</span> <span>test_size</span><span>=</span><span>0.20</span><span>)</span> <span>clf</span> <span>=</span> <span>DecisionTreeClassifier</span><span>()</span> <span>clf</span><span>.</span><span>fit</span><span>(</span><span>X_train</span><span>,</span> <span>y_train</span><span>)</span> <span>print</span><span>(</span><span>clf</span><span>.</span><span>feature_importances_</span><span>)</span></code></pre></div> <p>This results in <code>[ 0.18874114 0.34113588 0.47012298 ]</code>, meaning that all three features are somewhat important, with <em>density</em> accounting for almost 50% of the purity of a node—if you are unfamiliar with decision trees, the idea is to obtain nodes that are as “pure” as possible, meaning that there should be as few differences in class labels as possible. The <em>density</em> attribute appears to be important for splitting up impure nodes correctly.</p> <p>To get a better feeling of the feature space, let us briefly visualize it using <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">“principal component analysis”</a>:</p> <div><pre><code><span>clf</span> <span>=</span> <span>PCA</span><span>(</span><span>n_components</span><span>=</span><span>2</span><span>)</span> <span>X_</span> <span>=</span> <span>clf</span><span>.</span><span>fit_transform</span><span>(</span><span>X</span><span>)</span> <span>for</span> <span>label</span> <span>in</span> <span>set</span><span>(</span><span>labels</span><span>):</span> <span>idx</span> <span>=</span> <span>y</span><span>[</span><span>0</span><span>:,]</span> <span>==</span> <span>label</span> <span>plt</span><span>.</span><span>scatter</span><span>(</span><span>X_</span><span>[</span><span>idx</span><span>,</span> <span>0</span><span>],</span> <span>X_</span><span>[</span><span>idx</span><span>,</span> <span>1</span><span>],</span> <span>label</span><span>=</span><span>label</span><span>,</span> <span>alpha</span><span>=</span><span>0.25</span><span>)</span> <span>plt</span><span>.</span><span>legend</span><span>()</span> <span>plt</span><span>.</span><span>show</span><span>()</span></code></pre></div> <p>This results in the following image:</p> <figure><a href="http://bastian.rieck.me/images/decision_tree_features_pca.svg"> <img src="http://bastian.rieck.me/images/decision_tree_features_pca.svg" alt="A PCA visualization of the features used in the decision tree classifier"></a> </figure><p>Every dot in the image represents an individual graph, whereas distances in the plot roughly correspond to distances between features. The large amount of overlaps between graphs with different labels thus seems to suggest that the features are <em>insufficient</em> to separate the individual graphs. How does the decision tree manage to separate them, nonetheless? As it turns out, by creating a <em>very</em> deep and wide tree. To see this, let us export the decision tree classifier:</p> <div><pre><code><span>with</span> <span>open</span><span>(</span><span>"/tmp/clf.txt"</span><span>,</span> <span>"w"</span><span>)</span> <span>as</span> <span>f</span><span>:</span> <span>export_graphviz</span><span>(</span><span>clf</span><span>,</span> <span>f</span><span>)</span></code></pre></div> <p>Here is a small depiction of the resulting tree:</p> <figure><a href="http://bastian.rieck.me/images/decision_tree_complete.png"> <img src="http://bastian.rieck.me/images/decision_tree_complete.png" alt="A small version of the resulting decision tree"></a> </figure><p>There is also a <a href="http://bastian.rieck.me/blog/images/decision_tree_complete.svg">larger variant of this tree</a> for you to play around with. As you can see, the tree looks relatively complicated, so it is very much tailored to the given problem. That is not to say that the tree is necessarily suffering from <em>overfitting</em>—we explicitly used cross-validation to prevent this. What it <em>does</em> show is that the simple feature space with three features, while yielding very good classification results, is not detecting the true underlying structure of the problem. The rules generated by the decision tree are artificial in the sense that they do not help us understand what makes the two groups of graphs different from each other.</p> <h1 id="coda">Coda</h1> <p>So, where does this leave us? We saw that we were able to perform as well as state-of-the-art methods by merely picking a simple feature space, consisting of three features, and a simple decision tree classifier. I would say that <em>any</em> classifier with a higher complexity should yield significant improvements over this simple baseline, both in terms of average accuracy and in terms of standard deviation. Hence, the <a href="https://en.wikipedia.org/wiki/KISS_principle">KISS principle</a> in machine learning: start simple and only add more complex building blocks, i.e. models, if you have established a useful baseline.</p> <p>You can find both the code and the data in my <a href="https://github.com/Submanifold/TML">repository on topological machine learning</a>. To run the analysis described in this blog post, just execute</p> <pre><code>$ ./simple_network_analysis.py GML/????.gml Labels.txt </code></pre> <p>in the <code>REDDIT-BINARY</code> subfolder of the repository.</p> <p>That is all for now, until next time!</p> </div> </div><br>
<br>
<br>
<br>
<br>
<br>
via bastian.rieck.me http://bastian.rieck.me<br>
<br>
April 17, 2019 at 03:26PM
